Improvements to attribute table performance

nyalldawson opened the pull request:
This PR:

1. Implements a bunch of fixes to QgsVectorLayerCache which were preventing the cache being used in certain circumstances by the attribute table dialog
2. Ensures that the table does NOT load feature geometry by default (and adds tests to enforce this)
3. Moves partial responsibility for filtering attribute table to dual view master model. This avoids requiring the table to load ALL features when the table is set to just display selected or visible features. 

Together these changes dramatically improve the load time of the attribute table when working with large layers. And if you set the table to display selected features or visible features by default, then the benefits are enormous.

Not for merging yet - there's an issue with the dialog title & with occasional double-load of features on table load I need to address first.

On behalf of Faunalia, sponsored by ENEL

nyalldawson committed:
Clear existing cache when QgsVectorLayerCache is set to cache geom

Any features inside the cache must be cleared, because they
won't necessarily have the feature's geometry cached

On behalf of Faunalia, sponsored by ENEL

nyalldawson committed:
Small cleanup to QgsDualView - don't use cache to access layer

nyalldawson committed:
Flip some connects to new style

nyalldawson committed:
Ensure both attribute table cache & master model request respect
geometry fetching

If request needs geometry but cache isn't fetching it then cache
is bypassed. This is a performance hit, so ensure that cache
and request are always in sync wrt to fetching geoms.

On behalf of Faunalia, sponsored by ENEL

nyalldawson committed:
Add (failing) test to ensure that attribute table doesn't fetch
geometry by default

(because performance)

On behalf of Faunalia, sponsored by ENEL

nyalldawson committed:
Ensure that full cache flag is cleared when invalid

On behalf of Faunalia, sponsored by ENEL

nyalldawson committed:
Fix attribute table always fetches geometries for features

On behalf of Faunalia, sponsored by ENEL

nyalldawson committed:
Move partial responsibility for filtering attribute table to
dual view master model

This avoids requiring the table to load ALL features when the
table is set to just display selected or visible features

Dramatically improves load time of the attribute table when
working with large layers, so long as the table is set to
display selected features or visible features by default

On behalf of Faunalia, sponsored by ENEL

m-kuhn commented on src/core/qgsvectorlayercache.cpp:
```diff
@@ -58,7 +58,9 @@ int QgsVectorLayerCache::cacheSize()
 void QgsVectorLayerCache::setCacheGeometry( bool cacheGeometry )
 {
-  mCacheGeometry = cacheGeometry && mLayer->hasGeometryType();
+  bool shouldCache = cacheGeometry && mLayer->hasGeometryType();
```
Can you rename the local var to `shouldCacheGeom`, will save me a bit of thinking next time ;)

nyalldawson replied:
Fixed

m-kuhn commented on src/core/qgsvectorlayercache.cpp:
```diff
@@ -231,6 +237,7 @@ void QgsVectorLayerCache::attributeAdded( int field )
 {
   Q_UNUSED( field )
   mCachedAttributes.append( field );
+  mFullCache = false;
```
Hmm... `fullCache` is a flag set from externally for providers that do not allow to select by feature id and therefore require a full table scan for every single feature they request. I have a feeling that it's safer to re-scan the table when an attribute is added rather than disabling the full cache or we'll have a bad performance impact repeatedly rather than just once, no?

nyalldawson replied:
Thanks - fixed. I was under the mistaken impression that this was used to indicate that we'd filled the cache with all available features.

nyalldawson replied:
Actually I've reintroduced this fix, because it turns out it's valid. If the cache is invalidated for any reason (eg by adding an attribute to the table) and mFullCache is true, then there's no choice but to rebuild the entire cache from the table. This would be a performance killer, so my newly introduced fix makes it clear in the documentation that it's necessary to listen out for the invalidated signal and if required then handle rebuilding the full cache in a nice responsive way (eg with progress feedback). I've updated the attribute table dialog to suit.

m-kuhn commented:
Thanks for taking a look at this, @nyalldawson double-loading of the table really is a no go.

nyalldawson committed:
Revert 38f4a6f

nyalldawson commented:
@m-kuhn 
So just want to confirm that you're ok with moving the filtering to the model?

Another question: do you have any objections if i make "show selected features" show NO features when there's no selection? Current behaviour is to show all features. The issue with this approach is that you can't rely on a default setting of "show selected" to speed up the attribute table, because if you accidentally open the table with no selection (or clear the selection while the table is open) you'll get the whole table loading...

m-kuhn commented:
> So just want to confirm that you're ok with moving the filtering to the model?

It was a long time ago that I wrote this code.
Is there anything in particular that you think could be bad about this move?
In general I'm pro make attribute table faster and pro sending more filtered requests to the server.

> Another question: do you have any objections if i make "show selected features" show NO features when there's no selection? Current behaviour is to show all features. The issue with this approach is that you can't rely on a default setting of "show selected" to speed up the attribute table, because if you accidentally open the table with no selection (or clear the selection while the table is open) you'll get the whole table loading...

I am not completely sure but I think that this behavior was originally proposed by @pcav , I guess you have close contact to him in this project ;)

DelazJ commented:
@nyalldawson @pcav with this bug fix, does it mean that there would be no need to document the workaround described in https://github.com/qgis/QGIS-Documentation/issues/560?

nyalldawson committed:
Unvert "Ensure that full cache flag is cleared when invalid"

This is required - when the cache is invalidated it requires
a full rebuild (eg due to a new attribute being added) in order
to have complete information.

Since this could be a very lengthy process, it's not safe to
immediately rebuild the full cache. Instead, clear the full
cache flag and require users of this class to handle
responsive cache rebuilding by listening to the invalidated()
signal from the cache.

nyalldawson committed:
Avoid unnecessary reloads layer for attribute table

On behalf of Faunalia, sponsored by ENEL

nyalldawson committed:
Remove special handling for attribute table "show visible" initial state

This is no longer required as the show visible option now always
uses provider side filters

nyalldawson committed:
Use correct stats in attribute table dialog title

On behalf of Faunalia, sponsored by ENEL

nyalldawson commented:
> @nyalldawson @pcav with this bug fix, does it mean that there would be no need to document the workaround described in qgis/QGIS-Documentation#560?

Yes - kind of. That behavior is now redundant, however it's still necessary to set some mode except "show all features" as the default. If show all features is the default then you'll still get every feature loaded when the table is first opened.

nyalldawson commented:
@m-kuhn This should be ready to merge now. I'll leave it in master for testing for a couple of weeks before backporting to 2.18

nirvn commented:
@nyalldawson , any chance this will fix the issue with virtual field values not updating when a given feature's attributes are updated (http://hub.qgis.org/issues/15789)? I'm ask this as I see you mentioning caching improvements.

nyalldawson commented:
@nirvn 
> any chance this will fix the issue with virtual field values not updating when a given feature's attributes are updated (http://hub.qgis.org/issues/15789)? I'm ask this as I see you mentioning caching improvements.

No, sorry - no change there.

nirvn commented:
No worries then; the PR is very exciting as it is 😄

nyalldawson commented:
@m-kuhn any thoughts? am i OK to merge?

m-kuhn commented on src/core/qgsvectorlayercache.cpp:
```diff
@@ -267,6 +273,7 @@ void QgsVectorLayerCache::layerDeleted()
 void QgsVectorLayerCache::invalidate()
 {
   mCache.clear();
+  mFullCache = false;
```
Isn't this still silently disabling full cache with the potential of introducing performance impacts on invalidate?

nyalldawson replied:
Yes - but not silently. It emits the "invalidated" signal which is a flag to observers to rebuild a full cache (responsively!!) if required. That's what the attribute table does.

gioman commented:
> Not for merging yet - there's an issue with the dialog title & with occasional double-load of features on table load I need to address first.

Hi @nyalldawson ,
thanks for yours (and anyone involved) effort to improve this aspect in QGIS. I would like to ask you about the dataset/setup you used. I'm asking because I compiled this branch and I'm struggling to see improvements. My setup is a PostGIS 160k features (polygons) dataset on a server that is on the same local network (gigabit) as QGIS.

Actually when comparing to QGIS 2.18.4 when using "show all features", your branch take more than twice to load the table.

When using "show selected" or "show visible" I'm not able to see much difference in load time between this branch and 2.18.4/master.

I also noticed that when loading the layers with the "select at id" option disabled (for users this option would benefit a lot from a different wording) the scrolling of the table (when a lot of records are loaded) is much more fluid if compared with the same table loaded with the option checked (or using the browser). But when  "select at id" option is disabled the full table is always loaded, even when "show selected" or "show visible" are selected.

What about also @elpaso PR https://github.com/qgis/QGIS/pull/2518 ? the blocking attribute table when loading is still something that sets us behind compared to other gis packages.

cheers!

gioman commented:
Sorry @nyalldawson ,correction to my previous comment: I see (big) improvements when using the "show selected", but not when using "show visible".

gioman commented:
Another note: when using "show visible" when the whole dataset is visible loading the table is slower compared to the "show all" option: after the "*** features loaded" counter finishes it gets stuck for several seconds, this happens both on 2.18.4 and this branch, still on 2.18.4 is considerably faster.

gioman commented:
one final, maybe unrelated, note: rendering time in master/this branch is also considerably slower compared to 2.18.4.

gioman commented:
my notes/tests are maybe affected by this http://hub.qgis.org/issues/16239 I see a general degradation of performances between master and 2.18.4 when it comes to rendering and attribute table opening time, so I guess that it will be better to see the improvements from this PR once backported to 2.18(?).

nyalldawson commented:
> Actually when comparing to QGIS 2.18.4 when using "show all features", your branch take more than twice to load the table.

Discussion with @gioman leads me to believe that this 3.0 regression is unrelated - a simple python script looping through all features in a layer takes over twice as long. This is under investigation, but shouldn't block this work.

> What about also @elpaso PR #2518 ? the blocking attribute table when loading is still something that sets us behind compared to other gis packages.

Yes - that would also big a huge improvement! It needs considerable work to bring it back up to a merge-ready state for 3.0 however. @m-kuhn also has attribute table improvements inbound, so I think it's worth waiting for them and then revisiting the background loading option.

 > correction to my previous comment: I see (big) improvements when using the "show selected", but not when using "show visible".

That's good to hear - with this change the table should open near-instantly for smaller selections. Selecting all features and opening the table would be the same as before ;)

> Another note: when using "show visible" when the whole dataset is visible loading the table is slower compared to the "show all" option: after the "*** features loaded" counter finishes it gets stuck for several seconds, this happens both on 2.18.4 and this branch, still on 2.18.4 is considerably faster.

That's (somewhat) expected. When using show visible it does two things:
1. fetch all features from the current extent from the provider
2. checks to see what features are currently being rendered and filters for these

So if you are showing the entire layer, step 2 is what is causing the extra delay. It's necessary to avoid including features which may be hidden due to deselected categories, rule based render setup, etc. But it requires an iteration through the entire table to check this, so it's going to be slower.

gioman commented:
@nyalldawson thanks for your notes, much appreciated. Even if not related to this commit, could you give a word on the "select at id" thing, and if it is really expected that all features are loaded regardless the "show only..." options? Why loaded large tables are "fluid" to scroll with "select at id" and not without? Thanks in advance!

Note:
is it possible that the slow rendering regression has actually slipped in QGIS since a while now? see for example http://hub.qgis.org/issues/15752

nyalldawson commented:
@m-kuhn any objections if i merge?

m-kuhn commented:
Nope, go for it!

And thanks

nyalldawson committed:
Indentation

nyalldawson committed:
Update indentation, fix build

nyalldawson committed:
Update for QgsSettings
